app/models/search/bookmarkable_query.rb

```diff
@@ -98,9 +98,9 @@ def bookmarkable_query_text
   # field and sort by score).
   def sort
     if sort_column == "bookmarkable_date"
-      { revised_at: { order: sort_direction, unmapped_type: "date" } }
+      { revised_at: { order: sort_direction, unmapped_type: "date" }, _id: { order: sort_direction } }
```

Comment:
> Had to use _id for bookmarkables. ES discourages searching by _id, but I don't see other options.

I'm not sure if we should. ES has plans to eventually prevent using `_id` in sorting (https://github.com/elastic/elasticsearch/issues/64511). They added automatic tie breakers for `search_after`/point-in-time requests (https://github.com/elastic/elasticsearch/pull/68833), but for other use cases like ours, here's [what they recommend](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/paginate-search-results.html):

> If you cannot use a PIT, we recommend that you include a tiebreaker field in your `sort`. This tiebreaker field should contain a unique value for each document. If you don’t include a tiebreaker field, your paged results could miss or duplicate hits.

I'd wait for a second reviewer, but I think we should index bookmarkables with an "id" field (like we already did with works and the others). Because bookmarkables can be different object types with the same primary key, "id" should have the type for disambiguation, and we can probably reuse `document_id`:

https://github.com/otwcode/otwarchive/blob/1f8558452e6984a41ace47074324013d2a3454f8/app/models/search/bookmarkable_indexer.rb#L31-L33

If we use that approach, we'll need to reindex all works and bookmarkables (there are already rake tasks for that from #3987).

Reply:
I think red's comment makes sense.

Reply:
Thanks for explanations!
Added something, looks to be working.

- Named field "id" as suggested. Hope there won't be any clashes with bookmarks "id" in the same index as these are separate objects.
- Added mapping in the main bookmarks mapping as other places complained. Can't find in ES documents some values from the same mapping, like "creators", so not sure if there are not used anymore or if I put my code into incorrect place.
- I understand that rake jobs would be run after code deployment. But I am not sure why works must be reindexed if "_id" update touches only bookmarks index.

Reply:
> Can't find in ES documents some values from the same mapping, like "creators", so not sure if there are not used anymore or if I put my code into incorrect place.

The bookmarks index contains 4 types of documents: bookmarks + 3 types of bookmarkables (works, series, external works). The mapping is shared between all 4 types:

https://github.com/otwcode/otwarchive/blob/1f8558452e6984a41ace47074324013d2a3454f8/app/models/search/bookmarkable_indexer.rb#L11-L13

So `creators` is not used by bookmarks, but is used by the other 3.

> But I am not sure why works must be reindexed if "_id" update touches only bookmarks index.

Works appear in 2 indices: one for work search (using `WorkIndexer`/`WorkCreatorIndexer`) and one for bookmark search (as a bookmarkable type, using `BookmarkedWorkIndexer`). Because we change how bookmarkables are indexed, we need to reindex works, but only with `BookmarkedWorkIndexer`.

We'll run this rake task specifically:

https://github.com/otwcode/otwarchive/blob/1f8558452e6984a41ace47074324013d2a3454f8/lib/tasks/search.rake#L33-L38